src/main/daml/Daml/Finance/Instrument/Bond/Util.daml

```diff
@@ -51,11 +54,44 @@ includes : Schedule -> Schedule -> [Bool]
 includes scheduleA scheduleB =
   map (\a -> isSome $ find (\b -> b.adjustedEndDate == a.adjustedEndDate) scheduleB) scheduleA
 
+-- | Calculate the fixing date by using a day shift (if applicable) and then adjusting the date.
+calculateFixingDate : Date -> FixingDates -> HolidayCalendarData -> Int -> Date
+calculateFixingDate baseDate fixingDates fixingCalendars dayShift =
+  let
+    resetDateUnadjStart = case fixingDates.dayType of
+      None -> baseDate
+      Some Business -> addBusinessDays fixingCalendars dayShift baseDate
+      Some Calendar -> addDays baseDate dayShift
+  in
+    adjustDate fixingCalendars fixingDates.businessDayConvention resetDateUnadjStart
+
+-- | Apply a cap and a floor to a floating rate claim.
+capAndFloorRate : O -> Optional Decimal -> Optional Decimal -> O -> C -> Decimal -> C
+capAndFloorRate observedFloatingRate capRate floorRate dcf notionalAmount couponRate =
+  let
+    regularRate = observedFloatingRate + Const couponRate
+    dcfAndNotional = scale dcf notionalAmount
+    regularRateClaim = scale regularRate dcfAndNotional
```

Comment:
we are using a lot of `scale` nodes unnecessarily, I would add them only when we need them.

e.g. define `regularRateObservation = (observedFloatingRate + Const couponRate) * dcf`

and scale by `notionalAmount` afterwards